File: server/bin/main/com/example/reconix/server/service/FinanceService.kt
package com.example.reconix.server.service

import com.example.reconix.server.database.*
import com.example.reconix.shared.*
import kotlinx.datetime.Clock
import org.jetbrains.exposed.sql.*
import org.jetbrains.exposed.sql.SqlExpressionBuilder.eq
import org.jetbrains.exposed.sql.transactions.transaction

/**
 * Enhanced Finance Service for Dashboard and Analytics
 */
class FinanceService {

    /**
     * Get Dashboard Metrics for Finance Manager
     */
    fun getDashboardMetrics(): DashboardMetricsDTO {
        return transaction {
            // Get all invoices with their status
            val allInvoices = Invoices.selectAll().toList()

            val pendingInvoices = allInvoices.filter {
                it[Invoices.status] == InvoiceStatus.PENDING.name
            }
            val matchedInvoices = allInvoices.filter {
                it[Invoices.status] == InvoiceStatus.MATCHED.name
            }
            val mismatchedInvoices = allInvoices.filter {
                it[Invoices.status] == InvoiceStatus.MISMATCH.name
            }

            val totalPendingAmount = pendingInvoices.sumOf { it[Invoices.totalAmount] }
            val totalPayableAmount = matchedInvoices.sumOf { it[Invoices.totalAmount] }

            val totalValidated = matchedInvoices.size + mismatchedInvoices.size
            val matchRate = if (totalValidated > 0) {
                (matchedInvoices.size.toDouble() / totalValidated) * 100
            } else {
                0.0
            }

            // Get recent activity
            val recentActivity = InvoiceActions
                .innerJoin(Invoices)
                .innerJoin(Users)
                .selectAll()
                .orderBy(InvoiceActions.timestamp, SortOrder.DESC)
                .limit(10)
                .map { row ->
                    RecentActivityDTO(
                        id = row[InvoiceActions.id],
                        invoiceId = row[InvoiceActions.invoiceId],
                        vendorName = row[Invoices.vendorId], // Should join with vendor table
                        actionType = row[InvoiceActions.actionType],
                        performedBy = row[Users.fullName],
                        timestamp = row[InvoiceActions.timestamp],
                        amount = row[Invoices.totalAmount]
                    )
                }

            DashboardMetricsDTO(
                totalPendingInvoices = pendingInvoices.size,
                totalPendingAmount = totalPendingAmount,
                matchedInvoicesCount = matchedInvoices.size,
                mismatchedInvoicesCount = mismatchedInvoices.size,
                matchRate = matchRate,
                totalPayableAmount = totalPayableAmount,
                averageProcessingTime = "2.5 hours", // TODO: Calculate from actual data
                recentActivity = recentActivity
            )
        }
    }

    /**
     * Get Pending Invoices for Finance Manager Review
     */
    fun getPendingInvoices(): List<InvoiceListItemDTO> {
        return transaction {
            Invoices.selectAll()
                .where { Invoices.status eq InvoiceStatus.PENDING.name }
                .orderBy(Invoices.createdAt, SortOrder.DESC)
                .map { row ->
                    val invoiceId = row[Invoices.id]
                    val itemCount = InvoiceItems.selectAll()
                        .where { InvoiceItems.invoiceId eq invoiceId }
                        .count()
                        .toInt()

                    InvoiceListItemDTO(
                        id = invoiceId,
                        poId = row[Invoices.poId],
                        vendorName = row[Invoices.vendorId], // Should fetch from vendor table
                        totalAmount = row[Invoices.totalAmount],
                        status = InvoiceStatus.valueOf(row[Invoices.status]),
                        createdAt = row[Invoices.createdAt],
                        itemCount = itemCount,
                        matchPercentage = null
                    )
                }
        }
    }

    /**
     * Get Detailed 3-Way Match Data for an Invoice
     */
    fun getThreeWayMatch(invoiceId: String): ThreeWayMatchDTO? {
        return transaction {
            val invoiceRow = Invoices.selectAll()
                .where { Invoices.id eq invoiceId }
                .singleOrNull() ?: return@transaction null

            val poId = invoiceRow[Invoices.poId]

            // Get PO items
            val poItems = PurchaseOrderItems.selectAll()
                .where { PurchaseOrderItems.poId eq poId }
                .associate { row ->
                    row[PurchaseOrderItems.itemId] to Triple(
                        row[PurchaseOrderItems.itemName],
                        row[PurchaseOrderItems.quantity],
                        row[PurchaseOrderItems.unitPrice]
                    )
                }

            // Get GRN quantities
            val grnIds = Grns.selectAll()
                .where { Grns.poId eq poId }
                .map { it[Grns.id] }

            val grnQuantities = mutableMapOf<String, Int>()
            for (grnId in grnIds) {
                GrnItems.selectAll()
                    .where { GrnItems.grnId eq grnId }
                    .forEach { row ->
                        val itemId = row[GrnItems.itemId]
                        val qty = row[GrnItems.receivedQuantity]
                        grnQuantities[itemId] = grnQuantities.getOrDefault(itemId, 0) + qty
                    }
            }

            // Get invoice items
            val invoiceItems = InvoiceItems.selectAll()
                .where { InvoiceItems.invoiceId eq invoiceId }
                .toList()

            // Build detailed match results
            val matchDetails = invoiceItems.map { invRow ->
                val itemId = invRow[InvoiceItems.itemId]
                val (itemName, poQty, poPrice) = poItems[itemId]
                    ?: Triple("Unknown Item", 0, 0.0)
                val grnQty = grnQuantities[itemId] ?: 0
                val invQty = invRow[InvoiceItems.quantity]
                val invPrice = invRow[InvoiceItems.unitPrice]

                val quantityMatch = invQty <= grnQty
                val priceDiff = kotlin.math.abs(invPrice - poPrice)
                val priceMatch = priceDiff <= 0.05

                ValidationDetailDTO(
                    itemId = itemId,
                    itemName = itemName,
                    poQuantity = poQty,
                    grnQuantity = grnQty,
                    invoiceQuantity = invQty,
                    poPrice = poPrice,
                    invoicePrice = invPrice,
                    priceDifference = priceDiff,
                    quantityMatch = quantityMatch,
                    priceMatch = priceMatch,
                    overallMatch = quantityMatch && priceMatch
                )
            }

            val matchCount = matchDetails.count { it.overallMatch }
            val overallMatchPercentage = if (matchDetails.isNotEmpty()) {
                (matchCount.toDouble() / matchDetails.size) * 100
            } else {
                0.0
            }

            ThreeWayMatchDTO(
                invoiceId = invoiceId,
                poId = poId,
                vendorName = invoiceRow[Invoices.vendorId],
                invoiceDate = invoiceRow[Invoices.createdAt],
                totalAmount = invoiceRow[Invoices.totalAmount],
                status = InvoiceStatus.valueOf(invoiceRow[Invoices.status]),
                matchDetails = matchDetails,
                overallMatchPercentage = overallMatchPercentage,
                createdAt = invoiceRow[Invoices.createdAt],
                validatedAt = invoiceRow[Invoices.validatedAt]
            )
        }
    }

    /**
     * Approve an Invoice
     */
    fun approveInvoice(invoiceId: String, userId: Int, notes: String?): InvoiceActionResponse {
        return transaction {
            // Update invoice status
            val updated = Invoices.update({ Invoices.id eq invoiceId }) {
                it[status] = InvoiceStatus.MATCHED.name
                it[validatedAt] = Clock.System.now().toString()
            }

            if (updated == 0) {
                return@transaction InvoiceActionResponse(
                    success = false,
                    message = "Invoice not found",
                    invoiceId = invoiceId,
                    newStatus = InvoiceStatus.PENDING.name
                )
            }

            // Log the action
            InvoiceActions.insert {
                it[InvoiceActions.invoiceId] = invoiceId
                it[actionType] = "APPROVED"
                it[performedBy] = userId
                it[InvoiceActions.notes] = notes
                it[timestamp] = Clock.System.now().toString()
            }

            InvoiceActionResponse(
                success = true,
                message = "Invoice approved successfully",
                invoiceId = invoiceId,
                newStatus = InvoiceStatus.MATCHED.name
            )
        }
    }

    /**
     * Reject an Invoice
     */
    fun rejectInvoice(invoiceId: String, userId: Int, notes: String?): InvoiceActionResponse {
        return transaction {
            // Update invoice status
            val updated = Invoices.update({ Invoices.id eq invoiceId }) {
                it[status] = InvoiceStatus.MISMATCH.name
                it[validatedAt] = Clock.System.now().toString()
            }

            if (updated == 0) {
                return@transaction InvoiceActionResponse(
                    success = false,
                    message = "Invoice not found",
                    invoiceId = invoiceId,
                    newStatus = InvoiceStatus.PENDING.name
                )
            }

            // Log the action
            InvoiceActions.insert {
                it[InvoiceActions.invoiceId] = invoiceId
                it[actionType] = "REJECTED"
                it[performedBy] = userId
                it[InvoiceActions.notes] = notes
                it[timestamp] = Clock.System.now().toString()
            }

            InvoiceActionResponse(
                success = true,
                message = "Invoice rejected",
                invoiceId = invoiceId,
                newStatus = InvoiceStatus.MISMATCH.name
            )
        }
    }

    /**
     * Store Detailed Validation Logs
     */
    fun logValidationDetails(invoiceId: String, details: List<ValidationDetailDTO>) {
        transaction {
            for (detail in details) {
                ValidationLogs.insert {
                    it[ValidationLogs.invoiceId] = invoiceId
                    it[itemId] = detail.itemId
                    it[poQuantity] = detail.poQuantity
                    it[grnQuantity] = detail.grnQuantity
                    it[invoiceQuantity] = detail.invoiceQuantity
                    it[poPrice] = detail.poPrice
                    it[invoicePrice] = detail.invoicePrice
                    it[priceDifference] = detail.priceDifference
                    it[quantityMatch] = detail.quantityMatch
                    it[priceMatch] = detail.priceMatch
                    it[overallMatch] = detail.overallMatch
                    it[timestamp] = Clock.System.now().toString()
                }
            }
        }
    }
}

File: python_server/services/finance_service.py
"""
services/finance_service.py - Dashboard metrics, pending invoices, 3-way match, approve/reject
Equivalent to server/.../service/FinanceService.kt
"""
from datetime import datetime, timezone
from typing import List, Optional

from sqlalchemy import desc
from sqlalchemy.orm import Session

from models import (
    Grn, GrnItem, Invoice, InvoiceAction, InvoiceItem,
    PurchaseOrderItem, User, ValidationLog,
)
from schemas import (
    DashboardMetricsDTO, InvoiceActionResponse, InvoiceListItemDTO,
    InvoiceStatus, RecentActivityDTO, ThreeWayMatchDTO, ValidationDetailDTO,
)


def get_dashboard_metrics(db: Session) -> DashboardMetricsDTO:
    all_invoices = db.query(Invoice).all()
    pending = [i for i in all_invoices if i.status == InvoiceStatus.PENDING.value]
    matched = [i for i in all_invoices if i.status == InvoiceStatus.MATCHED.value]
    mismatched = [i for i in all_invoices if i.status == InvoiceStatus.MISMATCH.value]

    total_pending_amount = sum(i.total_amount for i in pending)
    total_payable_amount = sum(i.total_amount for i in matched)
    total_validated = len(matched) + len(mismatched)
    match_rate = (len(matched) / total_validated * 100) if total_validated > 0 else 0.0

    # Recent activity - join invoice_actions -> invoices -> users
    recent_rows = (
        db.query(InvoiceAction, Invoice, User)
        .join(Invoice, InvoiceAction.invoice_id == Invoice.id)
        .join(User, InvoiceAction.performed_by == User.id)
        .order_by(desc(InvoiceAction.timestamp))
        .limit(10)
        .all()
    )
    recent_activity = [
        RecentActivityDTO(
            id=action.id,
            invoiceId=action.invoice_id,
            vendorName=invoice.vendor_id,
            actionType=action.action_type,
            performedBy=user.full_name,
            timestamp=action.timestamp,
            amount=invoice.total_amount,
        )
        for action, invoice, user in recent_rows
    ]

    return DashboardMetricsDTO(
        totalPendingInvoices=len(pending),
        totalPendingAmount=total_pending_amount,
        matchedInvoicesCount=len(matched),
        mismatchedInvoicesCount=len(mismatched),
        matchRate=match_rate,
        totalPayableAmount=total_payable_amount,
        averageProcessingTime="2.5 hours",
        recentActivity=recent_activity,
    )


def get_pending_invoices(db: Session) -> List[InvoiceListItemDTO]:
    invoices = (
        db.query(Invoice)
        .filter(Invoice.status == InvoiceStatus.PENDING.value)
        .order_by(desc(Invoice.created_at))
        .all()
    )
    result = []
    for inv in invoices:
        item_count = db.query(InvoiceItem).filter(InvoiceItem.invoice_id == inv.id).count()
        result.append(InvoiceListItemDTO(
            id=inv.id,
            poId=inv.po_id or "",
            vendorName=inv.vendor_id,
            totalAmount=inv.total_amount,
            status=InvoiceStatus(inv.status),
            createdAt=inv.created_at,
            itemCount=item_count,
            matchPercentage=None,
            rejectionReason=inv.rejection_reason,
        ))
    return result


def get_three_way_match(db: Session, invoice_id: str) -> Optional[ThreeWayMatchDTO]:
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        return None

    po_id = invoice.po_id

    # PO items: {itemId -> (name, qty, price)}
    po_items = {
        row.item_id: (row.item_name, row.quantity, row.unit_price)
        for row in db.query(PurchaseOrderItem).filter(PurchaseOrderItem.po_id == po_id).all()
    }

    # GRN quantities
    grn_ids = [g.id for g in db.query(Grn).filter(Grn.po_id == po_id).all()]
    grn_quantities: dict[str, int] = {}
    for grn_id in grn_ids:
        for gi in db.query(GrnItem).filter(GrnItem.grn_id == grn_id).all():
            grn_quantities[gi.item_id] = grn_quantities.get(gi.item_id, 0) + gi.received_quantity

    invoice_items = db.query(InvoiceItem).filter(InvoiceItem.invoice_id == invoice_id).all()
    match_details: List[ValidationDetailDTO] = []
    for inv_item in invoice_items:
        item_id = inv_item.item_id
        po_data = po_items.get(item_id, ("Unknown Item", 0, 0.0))
        item_name, po_qty, po_price = po_data
        grn_qty = grn_quantities.get(item_id, 0)
        inv_qty = inv_item.quantity
        inv_price = inv_item.unit_price
        quantity_match = inv_qty <= grn_qty
        price_diff = abs(inv_price - po_price)
        price_match = price_diff <= 0.05
        match_details.append(ValidationDetailDTO(
            itemId=item_id,
            itemName=item_name,
            poQuantity=po_qty,
            grnQuantity=grn_qty,
            invoiceQuantity=inv_qty,
            poPrice=po_price,
            invoicePrice=inv_price,
            priceDifference=price_diff,
            quantityMatch=quantity_match,
            priceMatch=price_match,
            overallMatch=quantity_match and price_match,
        ))

    match_count = sum(1 for d in match_details if d.overallMatch)
    overall_pct = (match_count / len(match_details) * 100) if match_details else 0.0

    return ThreeWayMatchDTO(
        invoiceId=invoice_id,
        poId=po_id or "",
        vendorName=invoice.vendor_id,
        invoiceDate=invoice.created_at,
        totalAmount=invoice.total_amount,
        status=InvoiceStatus(invoice.status),
        matchDetails=match_details,
        overallMatchPercentage=overall_pct,
        createdAt=invoice.created_at,
        validatedAt=invoice.validated_at,
    )


def approve_invoice(db: Session, invoice_id: str, user_id: int, notes: Optional[str]) -> InvoiceActionResponse:
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        return InvoiceActionResponse(
            success=False,
            message="Invoice not found",
            invoiceId=invoice_id,
            newStatus=InvoiceStatus.PENDING.value,
        )
    now = datetime.now(timezone.utc).isoformat()
    invoice.status = InvoiceStatus.MATCHED.value
    invoice.validated_at = now
    db.add(InvoiceAction(
        invoice_id=invoice_id,
        action_type="APPROVED",
        performed_by=user_id,
        notes=notes,
        timestamp=now,
    ))
    db.commit()
    return InvoiceActionResponse(
        success=True,
        message="Invoice approved successfully",
        invoiceId=invoice_id,
        newStatus=InvoiceStatus.MATCHED.value,
    )


def reject_invoice(db: Session, invoice_id: str, user_id: int, notes: Optional[str]) -> InvoiceActionResponse:
    invoice = db.query(Invoice).filter(Invoice.id == invoice_id).first()
    if not invoice:
        return InvoiceActionResponse(
            success=False,
            message="Invoice not found",
            invoiceId=invoice_id,
            newStatus=InvoiceStatus.PENDING.value,
        )
    now = datetime.now(timezone.utc).isoformat()
    invoice.status = InvoiceStatus.MISMATCH.value
    invoice.validated_at = now
    invoice.rejection_reason = notes or "Rejected by finance manager"
    db.add(InvoiceAction(
        invoice_id=invoice_id,
        action_type="REJECTED",
        performed_by=user_id,
        notes=notes,
        timestamp=now,
    ))
    db.commit()
    return InvoiceActionResponse(
        success=True,
        message="Invoice rejected",
        invoiceId=invoice_id,
        newStatus=InvoiceStatus.MISMATCH.value,
    )


def log_validation_details(db: Session, invoice_id: str, details: List[ValidationDetailDTO]) -> None:
    now = datetime.now(timezone.utc).isoformat()
    for detail in details:
        db.add(ValidationLog(
            invoice_id=invoice_id,
            item_id=detail.itemId,
            po_quantity=detail.poQuantity,
            grn_quantity=detail.grnQuantity,
            invoice_quantity=detail.invoiceQuantity,
            po_price=detail.poPrice,
            invoice_price=detail.invoicePrice,
            price_difference=detail.priceDifference,
            quantity_match=detail.quantityMatch,
            price_match=detail.priceMatch,
            overall_match=detail.overallMatch,
            timestamp=now,
        ))
    db.commit()


def get_recent_activity(db: Session, limit: int = 20) -> List[RecentActivityDTO]:
    """
    Return the most recent invoice actions as an activity feed.
    Equivalent to the /dashboard/activity endpoint.
    """
    rows = (
        db.query(InvoiceAction, Invoice, User)
        .join(Invoice, InvoiceAction.invoice_id == Invoice.id)
        .join(User, InvoiceAction.performed_by == User.id)
        .order_by(desc(InvoiceAction.timestamp))
        .limit(limit)
        .all()
    )
    return [
        RecentActivityDTO(
            id=action.id,
            invoiceId=action.invoice_id,
            vendorName=invoice.vendor_id,
            actionType=action.action_type,
            performedBy=user.full_name,
            timestamp=action.timestamp,
            amount=invoice.total_amount,
        )
        for action, invoice, user in rows
    ]
